pkg/acpi/rsdp.go

```diff
@@ -0,0 +1,238 @@
+func (r *RSDP) AllData() []byte {
+	return r.data[:]
+}
+
+// TableData returns the RSDP table data as a []byte
+func (r *RSDP) TableData() []byte {
+	return r.data[36:]
+}
+
+// Sig returns the RSDP signature
+func (r *RSDP) Sig() string {
+	return string(r.data[:8])
+}
+
+// OEMID returns the RSDP OEMID
+func (r *RSDP) OEMID() string {
+	return string(r.data[9:15])
+}
+
+// OEMTableID returns the RSDP OEMTableID
+func (r *RSDP) OEMTableID() string {
+	return "rsdp?"
+}
+
+// Revision returns the RSDP revision, which
+// after 2002 should be >= 2
+func (r *RSDP) Revision() uint8 {
+	return r.revision
+}
+
+// OEMRevision returns the table OEMRevision.
+func (r *RSDP) OEMRevision() uint32 {
+	return 0
+}
+
+// CheckSum returns the table CheckSum.
+func (r *RSDP) CheckSum() uint8 {
+	return uint8(r.checksum)
+}
+
+// CreatorID returns the table CreatorID.
+func (r *RSDP) CreatorID() uint32 {
+	return uint32(0)
+}
+
+// CreatorRevision returns the table CreatorRevision.
+func (r *RSDP) CreatorRevision() uint32 {
+	return 0
+}
+
+// Base returns a base address or the [RX]SDT.
+// It will preferentially return the XSDT, but if that is
+// 0 it will return the RSDT address.
+func (r *RSDP) Base() int64 {
+	Debug("Base %v data len %d", r, len(r.data))
+	b := int64(binary.LittleEndian.Uint32(r.data[16:20]))
+	if b != 0 {
+		return b
+	}
+	return int64(binary.LittleEndian.Uint64(r.data[24:32]))
+}
+
+func readRSDP(base int64) (*RSDP, error) {
+	r := &RSDP{}
+	r.base = uint64(base)
+	for i := range r.data {
+		var d io.Uint8
+		if err := io.Read(base+int64(i), &d); err != nil {
+			return nil, err
+		}
+		r.data[i] = uint8(d)
+	}
+	return r, nil
+}
+
+func getRSDPEFI() (*RSDP, error) {
+	file, err := os.Open("/sys/firmware/efi/systab")
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	const (
+		acpi20 = "ACPI20="
+		acpi   = "ACPI="
+	)
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		start := ""
+		if strings.HasPrefix(line, acpi20) {
+			start = strings.TrimPrefix(line, acpi20)
+		}
+		if strings.HasPrefix(line, acpi) {
+			start = strings.TrimPrefix(line, acpi)
+		}
+		if start == "" {
+			continue
+		}
+		rsdp, err := strconv.ParseInt(start, 0, 64)
+		if err != nil {
+			continue
+		}
+		return readRSDP(rsdp)
+	}
+	if err := scanner.Err(); err != nil {
+		log.Printf("error while reading EFI systab: %v", err)
+	}
+	return nil, fmt.Errorf("invalid efi/systab file")
+}
+
+func num(n string, i int) (uint64, error) {
+	b, err := ioutil.ReadFile(fmt.Sprintf("/sys/firmware/memmap/%d/%s", i, n))
+	if err != nil {
+		return 0, err
+	}
+	start, err := strconv.ParseUint(string(b), 0, 64)
+	return start, err
+}
+
+// get RSDPmem is the option of last choice, it just grovels through
+// the e0000-ffff0 area, 16 bytes at a time, trying to find an RSDP.
+// These are well-known addresses for 20+ years.
+func getRSDPmem() (*RSDP, error) {
+	for base := int64(0xe0000); base < 0xffff0; base += 16 {
```

Comment:
You mentioned before that the RSDP can also be in the EBDA.

Reply:
I did not find its location in the EBDA all that helpful. For now, I'm letting that go.